expenses-manager-backend/src/main/java/com/endava/expensesmanager/service/impl/ExpenseServiceImpl.java

```diff
@@ -53,11 +56,44 @@ public ExpenseDto updateExpense(int id, ExpenseDto expenseDto) {
     public void deleteExpense(int id) {
         expenseRepository.deleteById(id);
     }
+
     @Override
-    public List<ExpenseDto> getExpensesByUserId(int userId) {
-        List<Expense> expenses = expenseRepository.findByUserId(userId);
-        return expenses.stream()
+    public List<ExpenseDto> getExpensesByUserId(int userId, LocalDateTime startDate, LocalDateTime endDate) {
+        if (!validateDates(startDate, endDate)) {
+            throw new BadRequestException();
+        }
+        if (startDate == null) {
+            return expenseRepository.findByUserId(userId).stream()
+                    .map(expenseMapper::expenseToExpenseDto)
+                    .toList();
+        }
+        if (endDate == null) {
+            endDate = LocalDateTime.now();
+        }
+        return expenseRepository.findAllByUserIdAndDateBetween(userId, startDate, endDate).stream()
                 .map(expenseMapper::expenseToExpenseDto)
                 .toList();
+
+    }
+
+    private boolean validateDates(LocalDateTime startDate, LocalDateTime endDate) {
+        if (startDate == null && endDate == null) {
+            return true;
+        }
+        if (startDate != null && endDate != null) {
+            return startDate.isBefore(endDate);
+        }
+        if (startDate != null) {
+            return startDate.isBefore(LocalDateTime.now());
+        }
+        return false;
+    }
+
+    @ResponseStatus (value = HttpStatus.BAD_REQUEST)
+    private static class BadRequestException extends RuntimeException {
```

Comment:
Add a GlobalExceptionHandler, follow these steps https://nikhilsukhani.medium.com/mastering-exception-handling-in-spring-boot-using-controlleradvice-and-exceptionhandler-e676b5dd62ed